src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ActionAttributeRouteModel.cs

```diff
@@ -3,11 +3,141 @@
+            {
+                // This logic seems arbitrary but there's a good reason for it.
+                // 
+                // When we build the controller level selectors, any metadata or action constraints
+                // that aren't IRouteTemplateProvider will be included in all selectors. So we 
+                // pick any selector and then grab all of the stuff that isn't IRouteTemplateProvider
+                // then we've found all of the items that aren't routes.
+                //
+                // This is fragile wrt application model customizing the data - but no one has
+                // run into an issue with this and its pretty esoteric.
+                additionalSelector = new SelectorModel(actionModel.Controller.Selectors.First());
+                additionalSelector.AttributeRouteModel = null;
+
+                for (var i = additionalSelector.ActionConstraints.Count - 1; i >= 0; i--)
+                {
+                    if (additionalSelector.ActionConstraints[i] is IRouteTemplateProvider)
+                    {
+                        additionalSelector.ActionConstraints.RemoveAt(i);
+                    }
+                }
+
+                for (var i = additionalSelector.EndpointMetadata.Count - 1; i >= 0; i--)
+                {
+                    if (additionalSelector.EndpointMetadata[i] is IRouteTemplateProvider)
+                    {
+                        additionalSelector.EndpointMetadata.RemoveAt(i);
+                    }
+                }
+            }
+
+            var actionConstraints = new List<IActionConstraintMetadata>();
+
+            foreach (var actionSelector in actionModel.Selectors)
+            {
+                var actionRouteModel = actionSelector.AttributeRouteModel;
+
+                // We check the action to see if the template allows combination behavior
+                // (It doesn't start with / or ~/) so that in the case where we have multiple
+                // [Route] attributes on the controller we don't end up creating multiple
+                if (actionRouteModel != null && actionRouteModel.IsAbsoluteTemplate)
+                {
+                    // We're overriding the routes from the controller, but any *unbound* constraints
+                    // still apply.
+                    var selector = new SelectorModel(actionSelector);
+
+                    selector.AttributeRouteModel = AttributeRouteModel.CombineAttributeRouteModel(
+                        left: null,
+                        right: actionRouteModel);
+
+                    AddActionConstraints(selector, additionalSelector?.ActionConstraints);
+                    AddEndpointMetadata(selector, additionalSelector?.EndpointMetadata);
+
+                    yield return selector;
+                }
+                else if (controllerSelectors.Count > 0)
+                {
+                    for (var i = 0; i < controllerSelectors.Count; i++)
```

Comment:
I actually fixed a bug that made it easier to run into errors than it should have been. I'll point out the test that changed later in the PR.

This code used to loop over `controllerModel.Selectors` instead of `controllerSelectors` - the difference is that the former includes *conventional selectors*, and if that happens then you will **definitely** run into a bug. It seems unintended and since it results in an error, I changed it.